Fix: Inconsistent ports in canonical URLs

htdat opened the pull request:
Fix #198 

## Description 

This PR addresses inconsistent port issues for get_current_url() and a few other things: 

- Simplify the way we construct the current URL and respect the `force_https_canonicals` option. This is just a string replacement rather than take care of checking and concatenating scheme, host, URI, and port. 
- Change this function from `private` to `public` so that it can be tested with PHPUnit. 
- Add tests for this function.

Though this seems quite a lot of changes, I think it's worthwhile as this PR simplifies things and has tests. 

## Tests

#### For the main issue #198 with the port number:

1. A simple way I found is to use this npm package https://developer.wordpress.org/block-editor/reference-guides/packages/packages-env/#quick-tldr-instructions
2. After that, visit the homepage, a post page, a few other pages (category, tag, author), and ensure that URLs are correct in the meta value `@context`.

#### For normal sites (without port number):

Basically, follow up step 2 above.

htdat committed:
Fix get_current_url with a port number

GaryJones commented on tests/all-test.php:
```diff
@@ -591,4 +591,87 @@ public function test_user_logged_in_multisite() {
+		// Test homepage
+		$this->go_to( '/' );
+		$res = self::$parsely->get_current_url();
+		self::assertSame( strpos( $res, $expected, 0 ), 0 );
```
Use a more specific assertion - https://phpunit.readthedocs.io/en/7.5/assertions.html#assertstringstartswith

GaryJones commented on tests/all-test.php:
```diff
@@ -591,4 +591,87 @@ public function test_user_logged_in_multisite() {
+
+		// Test a specific post
+		$post_array = $this->create_test_post_array();
+		$post       = $this->factory->post->create( $post_array );
+		$this->go_to( '/?p=' . $post );
+		$res = self::$parsely->get_current_url();
```
You're testing against a post, but the default first argument of the method is `nonpost`. Seems like this should be `post`, along with the `$post` ID (rename variable as $post_id for clarity). Without `post` as the first arg value, there's a whole section of the `get_current_url()` method that isn't being covered in the tests.

Example:

```suggestion
		$post_id    = $this->factory->post->create( $post_array );
		$this->go_to( '/?p=' . $post_id );
		$res = self::$parsely->get_current_url( 'post', $post_id );
```

GaryJones commented on wp-parsely.php:
```diff
@@ -1649,29 +1648,19 @@ private function get_current_url( $post = 'nonpost', $post_id = 0 ) {
+					: '';
 
-			return $canonical;
+			$url = site_url( $request_uri ); // Input var okay.
```
That comment isn't doing anything.

GaryJones commented on wp-parsely.php:
```diff
@@ -1649,29 +1648,19 @@ private function get_current_url( $post = 'nonpost', $post_id = 0 ) {
-			} elseif ( isset( $_SERVER['HTTP_HOST'], $_SERVER['REQUEST_URI'] ) ) { // Input var okay.
-				$canonical = $scheme . sanitize_text_field( wp_unslash( $_SERVER['HTTP_HOST'] ) ) . sanitize_text_field( wp_unslash( $_SERVER['REQUEST_URI'] ) ); // Input var okay.
-			}
+			$url = apply_filters( 'wp_parsely_permalink', $permalink, $post, $post_id );
```
Can this filter be made available to the other branch as well? Does it make to do that?

htdat replied:
I keep it as-is like what the original code is doing.
I've looked at this filter again. It is passing `$post_id`, which is not available in another conditional branch so we should not make it available outside of the current scope.

GaryJones commented on wp-parsely.php:
```diff
@@ -1631,12 +1631,11 @@ private function get_settings_url() {
-		$options = $this->get_options();
-		$scheme  = ( $options['force_https_canonicals'] ? 'https://' : 'http://' );
+	public function get_current_url( $post = 'nonpost', $post_id = 0 ) {
+		$url = '';
```
What does this assignment add?

htdat replied:
I do not have any strong opinion here. I just think it makes reading easier with a declaration somewhere. However, PHPStorm is also marking it's not necessary, then I am removing it.

GaryJones commented on tests/all-test.php:
```diff
@@ -591,4 +591,87 @@ public function test_user_logged_in_multisite() {
+				false,
+				'https://example.com:1234',
+				'http://example.com:1234',
+			),
```
How about mixing some of these up to test other strings, like `http://localhost:999` or `https://my-domain.local:1000/foo/bar`, since you're updating the site URL anyway?

htdat committed:
Updated the correct test case

Side-change: I've changed to `home_url` and update `home` option as we're dealing with the front-end links

htdat committed:
PHPUnit: Switch from assertSame to assertstringstartswith

htdat committed:
Remove unnecessary comment

htdat committed:
Remove an unnecessary assignment

htdat committed:
Cover more test cases: site has an address with a port number and it's installed under a folder

htdat commented on wp-parsely.php:
```diff
@@ -1649,29 +1646,19 @@ private function get_current_url( $post = 'nonpost', $post_id = 0 ) {
-				$canonical = $scheme . sanitize_text_field( wp_unslash( $_SERVER['HTTP_HOST'] ) ) . sanitize_text_field( wp_unslash( $_SERVER['REQUEST_URI'] ) ); // Input var okay.
-			}
+			$url = apply_filters( 'wp_parsely_permalink', $permalink, $post, $post_id );
+		} else {
+			$request_uri = isset( $_SERVER['REQUEST_URI'] )
+					? sanitize_text_field( wp_unslash( $_SERVER['REQUEST_URI'] ) )
+					: '';
 
-			return $canonical;
+			$url = home_url( $request_uri );
 		}
-		$page_url = site_url( null, $scheme );
```
Previously, this uses `site_url`. However, since we're dealing with the canonical links (front-end links), not WordPress file links, the correct usage should be `home_url`. 

I am also updating the relevant line in the test file to reflect this change too.

htdat commented:
Thanks @GaryJones :D 
I've updated this PR to address your valuable feedback. In this iteration, the most major change is from `site_url` to `home_url` as I explained above.

htdat commented:
I've just noticed we're having this issue too #174. It's directly related to this PR. 
I will continue researching and taking this issue into account.

GaryJones reviewed (approved):
Tests look good, changes look good, manual inspection of applied PR on local environment shows URLs displaying correctly 👍🏻

htdat committed:
Merge branch 'develop' into fix/get_current_url

Resolve the conflict with develop branch
